src/crimson/osd/pg.h

```diff
@@ -36,99 +38,398 @@ namespace ceph::os {
 
 class PG : public boost::intrusive_ref_counter<
   PG,
-  boost::thread_unsafe_counter>
+  boost::thread_unsafe_counter>,
+  PeeringState::PeeringListener,
+  DoutPrefixProvider
 {
   using ec_profile_t = std::map<std::string,std::string>;
-  using cached_map_t = boost::local_shared_ptr<OSDMap>;
+  using cached_map_t = boost::local_shared_ptr<const OSDMap>;
 
+  spg_t pgid;
+  pg_shard_t pg_whoami;
+  coll_t coll;
+  ceph::os::CyanStore::CollectionRef coll_ref;
+  ghobject_t pgmeta_oid;
 public:
   PG(spg_t pgid,
      pg_shard_t pg_shard,
      pg_pool_t&& pool,
      std::string&& name,
-     std::unique_ptr<PGBackend> backend,
      cached_map_t osdmap,
-     ceph::net::Messenger& msgr);
-
-  epoch_t get_osdmap_epoch() const;
-  const pg_info_t& get_info() const;
-  const pg_stat_t& get_stats() const;
-  void clear_state(uint64_t mask);
-  bool test_state(uint64_t mask) const;
-  void set_state(uint64_t mask);
-  const PastIntervals& get_past_intervals() const;
-  pg_shard_t get_primary() const;
-  bool is_primary() const;
-  bool is_acting(pg_shard_t pg_shard) const;
-  bool is_up(pg_shard_t pg_shard) const;
-  pg_shard_t get_whoami() const;
-  epoch_t get_last_peering_reset() const;
-  void update_last_peering_reset();
-  epoch_t get_need_up_thru() const;
-  void update_need_up_thru(const OSDMap* o = nullptr);
-
-  bool proc_replica_info(pg_shard_t from,
-			 const pg_info_t& pg_info,
-			 epoch_t send_epoch);
-  void proc_replica_log(pg_shard_t from,
-			const pg_info_t& pg_info,
-			const pg_log_t& pg_log,
-			const pg_missing_t& pg_missing);
-
-  using peer_info_t = std::map<pg_shard_t, pg_info_t>;
-  pg_shard_t find_best_info(const PG::peer_info_t& infos) const;
-  enum class choose_acting_t {
-    dont_change,
-    should_change,
-    pg_incomplete,
+     ceph::osd::ShardServices &shard_services,
+     ec_profile_t profile);
+
+  // EpochSource
+  epoch_t get_osdmap_epoch() const override final {
```

Comment:
nit, IMHO, `final` would suffice. and i think https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final explains the rationale well.

Reply:
Ok, will switch them all.